src/numerics/petsc_matrix.C

```diff
@@ -1201,7 +1202,82 @@ void PetscMatrix<T>::add (const T a_in, const SparseMatrix<T> & X_in)
 }
 
 
+template <typename T>
+void PetscMatrix<T>::matrix_matrix_mult (SparseMatrix<T> & X_in, SparseMatrix<T> & Y_out)
+{
+  libmesh_assert (this->initialized());
+
+   // sanity check.
+   libmesh_assert_equal_to (this->n(), X_in.m());
+   libmesh_assert_equal_to (this->m(), Y_out.m());
+   libmesh_assert_equal_to (X_in.n(), Y_out.n());
+
+   const PetscMatrix<T> * X = cast_ptr<const PetscMatrix<T> *> (&X_in);
+   PetscMatrix<T> * Y = cast_ptr<PetscMatrix<T> *> (&Y_out);
+
+   PetscErrorCode ierr=0;
+
+   // the matrix from which we copy the values has to be assembled/closed
+   libmesh_assert(X->closed());
+
+   semiparallel_only();
+
+   ierr = MatMatMult(_mat, X->_mat, MAT_INITIAL_MATRIX, PETSC_DEFAULT, &Y->_mat);
+   LIBMESH_CHKERR(ierr);
+
+}
+
+template <typename T>
+void PetscMatrix<T>::add_sparse_matrix (const SparseMatrix<T> & spm,
+                        const std::map<numeric_index_type, numeric_index_type> & row_ltog,
+                        const std::map<numeric_index_type, numeric_index_type> & col_ltog,
+                        const T scalar)
+{
+  libmesh_assert_equal_to(spm.m(), row_ltog.size());
+  libmesh_assert_equal_to(spm.n(), col_ltog.size());
+  // make sure matrix has larger size than spm
+  libmesh_assert_greater_equal(this->m(), spm.m());
+  libmesh_assert_greater_equal(this->n(),spm.n());
 
+  if(!this->closed())
+    this->close();
+
+  PetscErrorCode ierr=0;
+
+  const PetscMatrix<T> * pscm=cast_ptr<const PetscMatrix<T> *>(&spm);
+
+  PetscInt ncols=0;
+
+  const PetscInt *lcols;
+  const PetscScalar *vals;
+
+  std::vector<PetscInt> gcols;
+  std::vector<PetscScalar> values;
+
+  for (auto ltog: row_ltog)
+  {
+    PetscInt grow[]={static_cast<PetscInt>(ltog.second)}; // global row index
+
+    ierr = MatGetRow(pscm->_mat, static_cast<PetscInt>(ltog.first), &ncols, &lcols, &vals);
+    LIBMESH_CHKERR(ierr);
+    // get global indices (gcols) from lcols, increment values = vals*scalar
+    gcols.resize(ncols);
+    values.resize(ncols);
+    for (auto i : index_range(gcols))
+    {
+      gcols[i] = libmesh_map_find(col_ltog, lcols[i]);
+      values[i] = scalar*vals[i];
+    }
+    ierr=MatSetValues(_mat, 1, grow, ncols, gcols.data(), values.data(),  ADD_VALUES);
+    LIBMESH_CHKERR(ierr);
+    ierr  = MatRestoreRow(pscm->_mat, static_cast<PetscInt>(ltog.first), &ncols, &lcols, &vals);
+    LIBMESH_CHKERR(ierr);
+  }
+  ierr=MatAssemblyBegin(_mat, MAT_FINAL_ASSEMBLY);
+  LIBMESH_CHKERR(ierr);
+  ierr=MatAssemblyEnd(_mat, MAT_FINAL_ASSEMBLY);
+  LIBMESH_CHKERR(ierr);
```

Comment:
You do not want to call MatAssembly until you have done all `add_sparse_matrix`. For example,

```
J-> add_sparse_matrix(A1, rows1, cols1)
J-> add_sparse_matrix(A2, rows2, cols2)
.
..
J-> add_sparse_matrix(A100, rows100, cols100),

J->close()
```

That is, we will call `close` in moose.

Reply:
> Just introduce `enum`.

@fdkong let me make a separate ticket for this one, I can take care of it.

Reply:
> You do not want to call MatAssembly

This seems like an important thing to fix before merge...

Reply:
> You do not want to call MatAssembly until you have done all `add_sparse_matrix`. For example,
> 
> ```
> J-> add_sparse_matrix(A1, rows1, cols1)
> J-> add_sparse_matrix(A2, rows2, cols2)
> .
> ..
> J-> add_sparse_matrix(A100, rows100, cols100),
> 
> J->close()
> ```
> 
> That is, we will call `close` in moose.

I thought it would be safer to close the matrix here every time when finishing adding a sparse matrix to the original matrix (also, to prevent needing to close the matrix manually). Is it very inefficient to do so?

Reply:
We can not do that since the matrix memory will be shrunk after you call `MatAssembly`.  If you did this, there were two issues:

1) You would use much more memory than you should 

2) The code would be really slow because of many allocs.

Reply:
> We can not do that since the matrix memory will be shrunk after you call `MatAssembly`. If you did this, there were two issues:
> 
> 1. You would use much more memory than you should
> 2. The code would be really slow because of many allocs.

okay I see. In this case it makes sense to avoid calling `MatAssembly` many times by putting it here in this function. Thanks!

Reply:
> > Just introduce `enum`.
> 
> @fdkong let me make a separate ticket for this one, I can take care of it.

@jwpeterson thanks!